DropWizard Integration

valentinpurrucker opened the pull request:
Closes #21 

- Adds a wizard that lets the user choose betwenn different copying strategies when dropping directories in the explorer.

valentinpurrucker committed:
Add JavaFX Wizard, when user wants to drop a directory with content.

Task: #21

valentinpurrucker committed:
Add recursive copying functionality when dropping multiple directories.
Correct behavior when dropping empty directory.

Task: #21

valentinpurrucker committed:
Fixed missing title in Wizard Pages.
Replaced functional interfaces with lambdas.

valentinpurrucker committed:
Integrate two pages into one single page. Add page image and
description.

Task: #21

valentinpurrucker committed:
Merge branch 'team02-development' into
feature/task_21-drop_recursivly-to-explorer

Conflicts:
	de.tu_bs.cs.isf.e4cf.core/src/de/tu_bs/cs/isf/e4cf/core/stringtable/E4CEventTable.java
	de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/CustomTreeCell.java
	de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/controller/ProjectExplorerViewController.java
	de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/stringtable/FileTable.java

Task: #21

timbittner commented:
Seems to be working and behaving as expected. I tested:
- Dropping Multiple Folders and Files at once with each copy option.
- Dropping a single Folder
- Dropping a single Empty Folder
- Dropping Multiple Empty Folders
- Dropping a single File
- Dropping a few files
- Dropping a Folder containing only a single empty folder

timbittner commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/controller/ProjectExplorerViewController.java:
```diff
@@ -323,6 +340,19 @@ public void toggleView(@UIEventTopic(E4CEventTable.EVENT_VIEW_TOGGLE) Object o)
+	@Inject
+	@Optional
+	public void dropElements(@UIEventTopic(E4CEventTable.EVENT_DROP_ELEMENT_IN_EXPLORER) Object o) {
+		System.out.println("About to copy a directory");
```
Remove Debug print statement :)

timbittner commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/wizards/ImportDirectoryPage.java:
```diff
@@ -0,0 +1,89 @@
+
+	private CopyStrategy copyStrategy = CopyStrategy.EMPTY;
+
+	public ImportDirectoryPage(String pageName, Path path, IEclipseContext context, ImageDescriptor imgDescriptor) {
```
Javadoc

timbittner reviewed (commented):
Minimal Changes, DocString, Debug print.
While you're at it: try using eclipse formatting (STRG+SHIFT+F) on java files for nice and carefree formatting. (Except if that breaks your git and all files are then overwritten 😆  )

valentinpurrucker committed:
Changes due to PR

- remove print statements
- fix format
- add missing JavaDoc

Task: #21

Jooshua commented on de.tu_bs.cs.isf.e4cf.core/src/de/tu_bs/cs/isf/e4cf/core/stringtable/E4CEventTable.java:
```diff
@@ -1,17 +1,19 @@
-	public static final String EVENT_FILTER_CHANGED = "EVENT_FILTER_CHANGED";
-	public static final String EVENT_VIEW_TOGGLE = "EVENT_VIEW_TOGGLE";
-}
+package de.tu_bs.cs.isf.e4cf.core.stringtable;
+
+public class E4CEventTable {
+	/** 
+	 * Project Explorer
+	 */
+	public static final String EVENT_REFRESH_PROJECT_VIEWER = "EVENT_REFRESH_PROJECT_VIEWER"; 
+	public static final String SELECTION_CHANGED_EVENT ="SELECTION_CHANGED_EVENT";
+	public static final String NEW_IEC_PROJECT_EVENT ="NEW_PROJECT_EVENT";
+	
+	public static final String EVENT_DROP_ELEMENT_IN_EXPLORER = "DROP_ELEMENT_IN_EXPLORER";
+	
+	public static final String UPDATE_STATUS_BAR = "UPDATE_STATUS_BAR";
+
+	public static final String EVENT_RENAME_PROJECT_EXPLORER_ITEM = "EVENT_RENAME_PROJECT_EXPLORER_ITEM";
+	
+	public static final String EVENT_FILTER_CHANGED = "EVENT_FILTER_CHANGED";
+	public static final String EVENT_VIEW_TOGGLE = "EVENT_VIEW_TOGGLE";
+}
```
This git bug is really annoying -.-

timbittner replied:
But it seems to happen less and less :D

valentinpurrucker replied:
Yeah i'm sorry :D :D

Jooshua commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/CustomTreeCell.java:
```diff
@@ -53,6 +55,7 @@ public CustomTreeCell(WorkspaceFileSystem workspaceFileSystem, FileImageProvider
 
 			final Dragboard db = event.getDragboard();
 			boolean success = false;
+			List<java.io.File> directories = new ArrayList<File>();
```
We can do just `File`?

Jooshua commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/stringtable/FileTable.java:
```diff
@@ -1,25 +1,27 @@
+import java.nio.file.Paths;
+
+public class FileTable {
+	public static final String PROJECT_EXPLORER_VIEW_FXML = Paths.get("ui", "view", "ProjectExplorerView.fxml")
+			.toString();
+
+	public static final String COPY_DIRECTORY_PAGE_FXML = Paths.get("ui", "view", "CopyDirectoryView.fxml").toString();
```
See comment in the #37. Also wondering about the line breaking here

Jooshua reviewed (changes requested):
Looks good, but I have one thing that we have to discuss: Right now if we for example drag and drop a file, then we move it right? For directories now we always copy them right? Maybe we should introduce a switch between copying and moving in the wizard? Not sure though, maybe I am missing something...

timbittner commented:
> Looks good, but I have one thing that we have to discuss: Right now if we for example drag and drop a file, then we move it right? For directories now we always copy them right? Maybe we should introduce a switch between copying and moving in the wizard? Not sure though, maybe I am missing something...

You are talking about in-tree (ProjectExplorer) drag and drop right?
Dragging from Windows File Explorer should definetly always COPY.

I like @Jooshua's Idea of having a Little Checkbox titled "Move instead of Copy Files" or something like that. Depending on that check the wizards behavior changes. All other in-Tree DnD behavior is still working as expected (move).

That new checkbox needs to be disabled and set to copy mode when invoking the wizard from file system.  That is something that can be derived from the DragSource and I believe we already did something llike that in the TreeCellFactory.

@valipu Do you think this is an easy fix to implement or more challenging and should be its own issue instead of a PR change?

valentinpurrucker commented:
> > Looks good, but I have one thing that we have to discuss: Right now if we for example drag and drop a file, then we move it right? For directories now we always copy them right? Maybe we should introduce a switch between copying and moving in the wizard? Not sure though, maybe I am missing something...
> 
> You are talking about in-tree (ProjectExplorer) drag and drop right?
> Dragging from Windows File Explorer should definetly always COPY.
> 
> I like @Jooshua's Idea of having a Little Checkbox titled "Move instead of Copy Files" or something like that. Depending on that check the wizards behavior changes. All other in-Tree DnD behavior is still working as expected (move).
> 
> That new checkbox needs to be disabled and set to copy mode when invoking the wizard from file system. That is something that can be derived from the DragSource and I believe we already did something llike that in the TreeCellFactory.
> 
> @valipu Do you think this is an easy fix to implement or more challenging and should be its own issue instead of a PR change?

isn't it easier to just include the mode (copy or move) directly to the `DropElement` @timbittner  ? I think no user expects 

> > Looks good, but I have one thing that we have to discuss: Right now if we for example drag and drop a file, then we move it right? For directories now we always copy them right? Maybe we should introduce a switch between copying and moving in the wizard? Not sure though, maybe I am missing something...
> 
> You are talking about in-tree (ProjectExplorer) drag and drop right?
> Dragging from Windows File Explorer should definetly always COPY.
> 
> I like @Jooshua's Idea of having a Little Checkbox titled "Move instead of Copy Files" or something like that. Depending on that check the wizards behavior changes. All other in-Tree DnD behavior is still working as expected (move).
> 
> That new checkbox needs to be disabled and set to copy mode when invoking the wizard from file system. That is something that can be derived from the DragSource and I believe we already did something llike that in the TreeCellFactory.
> 
> @valipu Do you think this is an easy fix to implement or more challenging and should be its own issue instead of a PR change?

@timbittner i think we can includ it since it is not really much to change

valentinpurrucker committed:
Changes due to PR

- add option to choose the transfer mode when dropped
- move DropElement in own class

Task: #21

valentinpurrucker commented:
When i tested dropping a child directory with content to another directroy inside the tree i get a PermissionDeniedException but i don't think thats because of the implementation because i used to get it when we were dropping files in the tree too.

Just to be clear: the directory is still moved / copied correclty even when there is this AccessDeniedExc

valentinpurrucker committed:
Fix behavior when file is already in target.

Task: #21

timbittner commented:
Unfortunately while testing your latest comment I found another critical issue:

If I drag a folder onto its parent folder and select the 'move instead of copy' property the folder will be deleted leading to data loss.

Jooshua commented:
> 
> 
> Unfortunately while testing your latest comment I found another critical issue:
> 
> If I drag a folder onto its parent folder and select the 'move instead of copy' property the folder will be deleted leading to data loss.

Säd :/ I will rereview it after this there is another commit regarding this issue. Also you can now merge the development into this branch.

valentinpurrucker committed:
Fix in-tree dnd issue

Task: #21

valentinpurrucker committed:
Merge branch 'team02-development' into
feature/task_21-drop_recursivly-to-explorer

Conflicts:
	de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/stringtable/FileTable.java

valentinpurrucker commented:
Okay so this issue should be fixed and i hope (:D) that i didn't break anything else. But feel free to hit me with some more bugs :D @timbittner @Jooshua

timbittner commented:
> Okay so this issue should be fixed and i hope (:D) that i didn't break anything else. But feel free to hit me with some more bugs :D @timbittner @Jooshua

Sure 😆 
The original issue seems fixed but when I enable move while selecting 'shallow' or 'folder only' mode I still loose a lot of files.
That leads me to believe that we should enable move-mode only when performing 'deep'-move

timbittner commented:
Minor Detail: When trying to drop a file into a folder where another file with that name already exists an error dialog pops up, which is fine, but it does not contain any hints on what went wrong 😄

valentinpurrucker commented:
> > Okay so this issue should be fixed and i hope (:D) that i didn't break anything else. But feel free to hit me with some more bugs :D @timbittner @Jooshua
> 
> Sure 
> The original issue seems fixed but when I enable move while selecting 'shallow' or 'folder only' mode I still loose a lot of files.
> That leads me to believe that we should enable move-mode only when performing 'deep'-move

What do you mean by loosing files in shallow or create only folder? Does that happen when you have multiple files at once or just in general? Just curious 😜

Jooshua commented:
> 
> 
> > > Okay so this issue should be fixed and i hope (:D) that i didn't break anything else. But feel free to hit me with some more bugs :D @timbittner @Jooshua
> > 
> > 
> > Sure
> > The original issue seems fixed but when I enable move while selecting 'shallow' or 'folder only' mode I still loose a lot of files.
> > That leads me to believe that we should enable move-mode only when performing 'deep'-move
> 
> What do you mean by loosing files in shallow or create only folder? Does that happen when you have multiple files at once or just in general? Just curious 😜

How about some discord voice then? We are discussing this ticket atm...

valentinpurrucker committed:
Changed Copy/Move Behavior when dropping directories

Task: #21

timbittner commented on de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/CustomTreeCell.java:
```diff
@@ -53,6 +57,8 @@ public CustomTreeCell(WorkspaceFileSystem workspaceFileSystem, FileImageProvider
 			final Dragboard db = event.getDragboard();
 			boolean success = false;
+			DropMode dropMode = DropMode.COPY;
+			List<File> directories = new ArrayList<File>();
```
This file needs more comments around fields and if-statements to make it easier to understand whats going on.
The code in and of itself is fine though.

valentinpurrucker committed:
Fix in tree dnd when parent and child is selected.

- add missing docs

Task: #21

valentinpurrucker committed:
Add more java doc

Task: #21
